Added Image::setImageFromMemory and Image::setImageFromRawData

iUltimateLP opened the pull request:
Adds support to set the image resource a brls::Image view shows from memory using nanovg's nvgCreateImageMem.

iUltimateLP committed:
Added Image::setImageFromMemory

natinusala commented on library/lib/views/image.cpp:
```diff
@@ -280,6 +281,34 @@ void Image::setImageFromFile(std::string path)
+        std::stringstream addr;
+        addr << (void*)data;
+
+        fatal("Cannot load image from memory (" + std::to_string(numData) + " bytes at 0x" + addr.str() + ")");
```
Why do you need a stringstream for that?

iUltimateLP replied:
To convert the address of the pointer to a std:: string (it needs to be a std::string so I can use the operator+). If there's an easier way though, let me know!

natinusala replied:
I don't think that's worth just for an error message IMO

natinusala commented:
Hi, thanks for the PR!

This is a duplicate of #159 but I can merge whichever PR is finished first.

Do you think you could add a way to test / showcase this in the demo?

iUltimateLP commented:
Ah didn't see the other PR :)

As for a demo case, I'd implement a new view element "DynamicImage" or something which generates a random Bitmap and renders that!

iUltimateLP committed:
Don't need address in error message

iUltimateLP committed:
Changed nvgCreateImageMem to nvgCreateImageRGBA and implemented a small demo

iUltimateLP commented:
Had some nasty bugs with memory allocation, which is why I switched from `nvgCreateImageMem` (which was pretty unnessecary anyway since it needed a whole image "file" in memory rather than just a raw rgba array) to `nvgCreateImageRGBA`. Also implemented a little example in `demo/dynamic_image` and `views/dynamicImage.xml`. If you click on the image, it will populate with a gradient - let me know what you think!

iUltimateLP committed:
Differentiated between setImageFromMemory and setImageFromRawData, as the first still might be useful for e.g. image files requested from the web

natinusala commented:
This is absolutely not the same, and if I were to choose I would keep the "from mem" instead of "from RGBA" method.

The "from mem" method is used to load JPG / PNG files from memory, and is actually currently used in the sys-clk manager to display game thumbnails. Any homebrew dealing with images on HOS would need it too (homebrew menu, qlaunch replacement, album replacement...).

I understand that it's hard to generate a random PNG in memory, so how about adding both methods and only making an example for thr RGBA one ?

iUltimateLP commented:
See my latest commit, I had the same thought :P

natinusala commented on demo/dynamic_image.cpp:
```diff
@@ -0,0 +1,74 @@
+/*
+    Copyright 2019 WerWolv
```
are you sure of the copyright ?

iUltimateLP replied:
Never worked with PRs and licensing, should it just be `Copyright 2020-2021 natinusala`?

natinusala replied:
No, the copyright is for the person who wrote the file so that would be 2021 [your name]!

natinusala commented on demo/dynamic_image.cpp:
```diff
@@ -0,0 +1,74 @@
+
+#include "dynamic_image.hpp"
+
+DynamicImage::DynamicImage()
```
I would not call it "dynamic image", I don't find it very explicit

How about something like "RandomNoiseImage"? Or RandomRGBAImage?

natinusala commented on library/include/borealis/views/image.hpp:
```diff
@@ -95,6 +95,23 @@ class Image : public View
+     * 
+     * See the example implementation dynamic_image in the demo.
+     */
+    void setImageFromRawData(const unsigned char* data, int width, int height);
```
"raw data" can be anything: RGB, RGBA, YUV, ABRG...

I would call it "setImageFromRGBA" rather than "raw data", and most importantly specify the _full_ pixel format in the documentation (RGBA8888 if I'm not mistaken?). This is better than asking for a "width * height * 4" array.

natinusala commented on library/lib/views/image.cpp:
```diff
@@ -280,6 +280,53 @@ void Image::setImageFromFile(std::string path)
+
+    if (this->texture == 0)
+    {
+        brls::Logger::error("Cannot load image from memory (" + std::to_string(width) + "x" + std::to_string(height) + ")");
```
wrong error string

natinusala commented on demo/dynamic_image.cpp:
```diff
@@ -0,0 +1,74 @@
+brls::View* DynamicImage::create()
+{
+    // Called by the XML engine to create a new DynamicImage
+    return new DynamicImage();
```
file missing from the demo meson file

natinusala commented on resources/xml/views/dynamic_image.xml:
```diff
@@ -0,0 +1,20 @@
+        width="auto"
+        height="auto"
+        horizontalAlign="center"
+        text="Click the image!" />
```
Please use i18n, and I think it would be better to have "Randomly generated RGB (click to refresh)" as a caption

natinusala commented on demo/dynamic_image.cpp:
```diff
@@ -0,0 +1,74 @@
+    size_t bufferSize = height * width * 4; // 4 bytes per pixel (RGBA)
+    unsigned char* imageData = (unsigned char*)calloc(1, bufferSize);
+
+    for (int y = 0; y < height; y++)
```
Do you think you could randomize it?

And also make it so that it's "populated" when the view is created, so that it doesn't start all white?

iUltimateLP replied:
Trying to run the image generation in the RandomRGBAImage constructor causes a crash, I'm pretty sure it's because this->image->getHeight() / width() returns 0, hence it tries to allocate 0 bytes. At what stage is it safe to get the dimensions?

natinusala replied:
You don't need the image width or height, you can just generate an image on a 256x256 grid and borealis will upscale it

iUltimateLP committed:
Renamed DynamicImage to RandomRGBAImage, added random image generation for demo, made label using i18n, included in meson.build

iUltimateLP committed:
Hardcode the size of the random rgba_image to 256x256

iUltimateLP committed:
Fixed "narrow conversion" warning

natinusala commented:
This is great! I feel like the example image is a little bit too big and too close to the other images (it's missing some padding).

Could you maybe make it a rectangle, like the "stretched" image example? And add some top padding. Thanks!

iUltimateLP committed:
Added padding and changed size of random rgba image

iUltimateLP commented:
Sure, here you go :)

iUltimateLP commented:
Any update @natinusala? :)

EmmmaTech commented on library/lib/views/image.cpp:
```diff
@@ -280,6 +280,53 @@ void Image::setImageFromFile(std::string path)
+    this->invalidate();   
+}
+
+void Image::setImageFromRGBA(const unsigned char* data, int width, int height)
```
Why is this function only `const unsigned char *` and the other isn't? Is there a reason for this?

iUltimateLP replied:
Absolutely right, can definitely be made const.

EmmmaTech reviewed (changes requested):
Since natinusala is busy, I decided to look at this PR myself. Here is one change you should consider making

iUltimateLP committed:
Made Image::setImageFromMemory's data buffer const too

iUltimateLP committed:
nanovg expects a non-const data buffer in Image::setImageFromMemory but doesn't alter the data, so we don't have to expose the non-constness to borealis
